pkg/providers/instance/types.go

```diff
@@ -16,81 +16,188 @@ package instance
+		Tags:         tags,
+	}))
+	return resolved
+}
+
+type CreateFleetInputBuilder struct {
+	capacityType          string
+	tagSpecifications     []ec2types.TagSpecification
+	launchTemplateConfigs []ec2types.FleetLaunchTemplateConfigRequest
+
+	contextID               *string
+	capacityReservationType v1.CapacityReservationType
+}
+
+func NewCreateFleetInputBuilder(capacityType string, tags map[string]string, launchTemplateConfigs []ec2types.FleetLaunchTemplateConfigRequest) *CreateFleetInputBuilder {
+	var taggedResources = []ec2types.ResourceType{
+		ec2types.ResourceTypeInstance,
+		ec2types.ResourceTypeVolume,
+		ec2types.ResourceTypeFleet,
+	}
+	return &CreateFleetInputBuilder{
+		capacityType: capacityType,
+		tagSpecifications: lo.Map(taggedResources, func(resource ec2types.ResourceType, _ int) ec2types.TagSpecification {
+			return ec2types.TagSpecification{ResourceType: resource, Tags: utils.EC2MergeTags(tags)}
+		}),
+		launchTemplateConfigs: launchTemplateConfigs,
+	}
+}
+
+func (b *CreateFleetInputBuilder) WithContextID(contextID string) *CreateFleetInputBuilder {
+	b.contextID = &contextID
+	return b
+}
+
+func (b *CreateFleetInputBuilder) WithCapacityReservationType(crt v1.CapacityReservationType) *CreateFleetInputBuilder {
+	if b.capacityType != karpv1.CapacityTypeReserved {
+		panic("can not specify capacity reservation type when capacity type is not reserved")
+	}
+	b.capacityReservationType = crt
+	return b
+}
+
+func (b *CreateFleetInputBuilder) defaultTargetCapacityType() ec2types.DefaultTargetCapacityType {
+	switch b.capacityType {
+	case karpv1.CapacityTypeReserved:
+		if b.capacityReservationType == v1.CapacityReservationTypeCapacityBlock {
+			return ec2types.DefaultTargetCapacityTypeCapacityBlock
+		} else {
+			return ec2types.DefaultTargetCapacityTypeOnDemand
+		}
+	case karpv1.CapacityTypeOnDemand, karpv1.CapacityTypeSpot:
+		return ec2types.DefaultTargetCapacityType(b.capacityType)
+	}
+	panic(fmt.Sprintf("invalid capacity type %q provided to create fleet input", b.capacityType))
+}
+
+func (b *CreateFleetInputBuilder) Build() *ec2.CreateFleetInput {
+	input := &ec2.CreateFleetInput{
+		Type:                  ec2types.FleetTypeInstant,
+		Context:               b.contextID,
+		LaunchTemplateConfigs: b.launchTemplateConfigs,
+		TargetCapacitySpecification: &ec2types.TargetCapacitySpecificationRequest{
+			DefaultTargetCapacityType: b.defaultTargetCapacityType(),
+			TotalTargetCapacity:       lo.ToPtr[int32](1),
+		},
+		TagSpecifications: b.tagSpecifications,
+	}
+	if b.capacityType == karpv1.CapacityTypeSpot {
+		input.SpotOptions = &ec2types.SpotOptionsRequest{
+			AllocationStrategy: ec2types.SpotAllocationStrategyPriceCapacityOptimized,
+		}
+	} else if b.capacityReservationType != v1.CapacityReservationTypeCapacityBlock {
+		input.OnDemandOptions = &ec2types.OnDemandOptionsRequest{
+			AllocationStrategy: ec2types.FleetOnDemandAllocationStrategyLowestPrice,
+		}
```

Comment:
If we provided `AllocationStrategy` for capacity blocks would it just be dropped? If so, can we just make this the default?

Reply:
CreateFleet actually fails if you provide an allocation strategy with capacity blocks, I'm really not sure why. I'd rather just leave it as the default, but at the moment this check is required.